core/src/test/java/org/elasticsearch/indices/recovery/IndexPrimaryRelocationIT.java

```diff
@@ -0,0 +1,234 @@
+    private String nodeBId = null;
+
+    private IndexShard nodeAShard = null;
+
+    private BlockClusterStateProcessing blockClusterStateProcessingOnNodeA;
+    private BlockClusterStateProcessing blockClusterStateProcessingOnNodeB;
+    private ServiceDisruptionScheme blockClusterStateProcessingOnBothNodes;
+
+    @Override
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
+
+        logger.info("--> start master node");
+        internalCluster().startMasterOnlyNode(Settings.EMPTY);
+        logger.info("--> start node A");
+        nodeA = internalCluster().startDataOnlyNode(Settings.EMPTY);
+
+        createAndPopulateIndex(INDEX_NAME, 1, SHARD_COUNT, REPLICA_COUNT);
+
+        logger.info("--> start node B");
+        nodeB = internalCluster().startDataOnlyNode(Settings.EMPTY);
+        ensureGreen();
+
+        client().admin().cluster().prepareState().get().getState().getNodes().forEach(node -> {
+            if (node.getName().equals(nodeA)) {
+                nodeAId = node.getId();
+            }
+            if (node.getName().equals(nodeB)) {
+                nodeBId = node.getId();
+            }
+        });
+        assertThat(nodeAId, notNullValue());
+        assertThat(nodeBId, notNullValue());
+
+        IndicesService indicesService = internalCluster().getInstance(IndicesService.class, nodeA);
+        final IndexService indexService = indicesService.indexServiceSafe(INDEX_NAME);
+        nodeAShard = indexService.getShard(0);
+
+        blockClusterStateProcessingOnNodeA = new BlockClusterStateProcessing(nodeA, random());
+        blockClusterStateProcessingOnNodeB = new BlockClusterStateProcessing(nodeB, random());
+
+        blockClusterStateProcessingOnBothNodes = new MultiServiceDisruptionScheme(blockClusterStateProcessingOnNodeA, blockClusterStateProcessingOnNodeB);
+    }
+
+
+    /**
+     * Tests two scenarios:
+     *
+     * Scenario 1:
+     *
+     * source node and target node on cluster state before relocation target is marked as started and shard on source node not yet marked as
+     * RELOCATED. This means that source node knows it is active primary but not yet relocated and target node knows it is primary and
+     * relocation target. Index requests to source node are indexed on source node and replicated to target node. Index requests to target
+     * node are rerouted to source node.
+     *
+     * Scenario 2:
+     *
+     * source node and target node on cluster state before relocation target is marked as started and shard on source node marked as
+     * RELOCATED. This means that source node knows it is relocated and target node knows it is a primary relocation target. Index
+     * requests to source node are sent in primary phase to target node and replicated back to source node. Index requests to
+     * target are rerouted back to source node.
+     */
+    public void testPrimaryRecoveryBothNodesOnOldClusterState() throws Exception {
```

Comment:
can we replace these test with a simpler to understand test, paying the price of things being less targeted? experience have shown that this type of tests are very hard to maintain and often don't reproduce exactly what was intended anyway (because it's so hard)..